process/block/alteredaccounts/alteredAccountsProvider_test.go

```diff
@@ -0,0 +1,430 @@
+				SndAddr: []byte("shard1 addr - tx3"),
+				RcvAddr: []byte("shard0 addr - tx3"),
+			},
+			"hash4": &transaction.Transaction{ // cross-shard, no address in shard 0
+				SndAddr: []byte("shard2 addr - tx4"),
+				RcvAddr: []byte("shard2 addr - tx3"),
+			},
+		},
+	})
+	require.NoError(t, err)
+	require.Equal(t, 5, len(res))
+
+	for key := range res {
+		decodedKey, _ := args.AddressConverter.Decode(key)
+		require.True(t, strings.HasPrefix(string(decodedKey), "shard0"))
+	}
+}
+
+func testExtractAlteredAccountsFromPoolTrieDataChecks(t *testing.T) {
+	t.Parallel()
+
+	receiverInSelfShard := "receiver in shard 1"
+	expectedBalance := big.NewInt(37)
+	args := getMockArgs()
+	args.ShardCoordinator = &testscommon.ShardsCoordinatorMock{
+		ComputeIdCalled: func(address []byte) uint32 {
+			if strings.HasPrefix(string(address), "sender") {
+				return 0
+			}
+
+			return 1
+		},
+		SelfIDCalled: func() uint32 {
+			return 1
+		},
+	}
+	args.AccountsDB = &state.AccountsStub{
+		LoadAccountCalled: func(container []byte) (vmcommon.AccountHandler, error) {
+			return &state.UserAccountStub{
+				Balance: expectedBalance,
+			}, nil
+		},
+	}
+	aap, _ := NewAlteredAccountsProvider(args)
+
+	res, err := aap.ExtractAlteredAccountsFromPool(&indexer.Pool{
+		Txs: map[string]data.TransactionHandler{
+			"hash0": &transaction.Transaction{
+				SndAddr: []byte("sender in shard 0"),
+				RcvAddr: []byte(receiverInSelfShard),
+			},
+		},
+	})
+	require.NoError(t, err)
+	require.Equal(t, 1, len(res))
+
+	expectedAddressKey := args.AddressConverter.Encode([]byte(receiverInSelfShard))
+	actualAccount, found := res[expectedAddressKey]
+	require.True(t, found)
+	require.Equal(t, expectedAddressKey, actualAccount.Address)
+	require.Equal(t, expectedBalance.String(), actualAccount.Balance)
+}
+
+func testExtractAlteredAccountsFromPoolShouldIncludeESDT(t *testing.T) {
+	t.Parallel()
+
+	expectedToken := esdt.ESDigitalToken{
+		Value: big.NewInt(37),
+	}
+	args := getMockArgs()
+	args.AccountsDB = &state.AccountsStub{
+		LoadAccountCalled: func(_ []byte) (vmcommon.AccountHandler, error) {
+			return &state.UserAccountStub{
+				RetrieveValueFromDataTrieTrackerCalled: func(_ []byte) ([]byte, error) {
+					tokenBytes, _ := args.Marshalizer.Marshal(expectedToken)
+					return tokenBytes, nil
+				},
+			}, nil
+		},
+	}
+	aap, _ := NewAlteredAccountsProvider(args)
+
+	res, err := aap.ExtractAlteredAccountsFromPool(&indexer.Pool{
+		Logs: []*data.LogData{
+			{
+				LogHandler: &transaction.Log{
+					Address: []byte("addr"),
+					Events: []*transaction.Event {
+						{
+							Address: []byte("addr"),
+							Identifier: []byte(core.BuiltInFunctionESDTTransfer),
+							Topics: [][]byte{
+								[]byte("token0"),
+							},
+						},
+					},
+				},
+			},
+		},
+	})
+	require.NoError(t, err)
+
+	encodedAddr := hex.EncodeToString([]byte("addr"))
+	require.Equal(t, expectedToken.Value.String(), res[encodedAddr].Tokens[0].Balance)
```

Comment:
Perhaps create an expected `AccountTokenData` object and compare it with `res[encodedAddr].Tokens[0]` so you can be sure that all fields are the same (e.g.: esdt name = `token0`)

Reply:
done